src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,143 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+
+    private int size;
+    private Node<T> first;
+    private Node<T> lastElement;
+
+    private void firstElement(T value) {
+        Node<T> newElement = new Node<>();
+        newElement.value = value;
+        first = newElement;
+        lastElement = newElement;
+        size++;
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        if (size == 0) {
+            firstElement(value);
+        } else {
+            Node<T> newElement = new Node<>();
+            newElement.value = value;
+            newElement.prev = lastElement;
+            lastElement.next = newElement;
+            if (lastElement.prev == null) {
+                first.next = newElement;
+            }
+            lastElement = newElement;
+            size++;
+        }
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException(index);
+        }
+        if (size == 0 || index == size) {
+            add(value);
+        } else {
+            Node<T> newElement = new Node<>();
+            newElement.value = value;
+            newElement.next = getElement(index);
+            getElement(index).prev = newElement;
 
+            if (index > 0) {
+                getElement(index - 1).next = newElement;
+                newElement.prev = getElement(index - 1);
+            } else {
+                first = newElement;
+            }
+            size++;
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (int i = 0; i < list.size(); i++) {
+            add((T) list.get(i));
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getElement(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T oldItem = getElement(index).value;
+        getElement(index).value = value;
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T removedItemValue = getElement(index).value;
+
+        if (size == 1) {
+            first = null;
+            lastElement = null;
+        } else {
+            if (getElement(index).next == null) {
+                lastElement = getElement(index - 1);
```

Comment:
instead of calling getElement() method each time (that's whole loop execution) let's save method result into temp variable